File: backend-go/internal/domain/engine/versioning_test.go
package engine

import (
	"errors"
	"reflect"
	"testing"

	"github.com/tiagofur/muebles-backend/internal/domain"
)

// TestStructureRevisionNumber_LegacyDefaults verifies that revision 0
// (missing in legacy data) is treated as DefaultStructureRevision.
func TestStructureRevisionNumber_LegacyDefaults(t *testing.T) {
	if got := StructureRevisionNumber(domain.Structure{}); got != DefaultStructureRevision {
		t.Errorf("legacy revision: got %d want %d", got, DefaultStructureRevision)
	}
	if got := StructureRevisionNumber(domain.Structure{Revision: 5}); got != 5 {
		t.Errorf("explicit revision: got %d want 5", got)
	}
}

// TestBumpStructureRevision increments revision and prepends the snapshot of
// the previous state (newest-first), mirroring the TS helper.
func TestBumpStructureRevision(t *testing.T) {
	current := domain.Structure{
		ID: "st-1", Code: "OLD", Name: "Old",
		Revision: 2,
		Components: []domain.ComponentInstance{{ComponentID: "c1", Quantity: 1}},
		History: []domain.StructureRevision{
			{Revision: 1, Code: "v1"},
		},
	}
	draft := domain.Structure{
		Code: "NEW", Name: "New",
		Components: []domain.ComponentInstance{{ComponentID: "c1", Quantity: 3}},
		// Draft smuggling attempts — must be ignored (parity with TS).
		Revision: 999, ID: "smuggled-id",
	}

	got := BumpStructureRevision(current, draft)
	if got.ID != "st-1" {
		t.Errorf("ID: got %q want st-1 (draft smuggle ignored)", got.ID)
	}
	if got.Revision != 3 {
		t.Errorf("Revision: got %d want 3", got.Revision)
	}
	if got.Code != "NEW" || got.Name != "New" {
		t.Errorf("draft fields not applied: code=%q name=%q", got.Code, got.Name)
	}
	if len(got.History) != 2 {
		t.Fatalf("History len: got %d want 2", len(got.History))
	}
	if got.History[0].Revision != 2 || got.History[0].Code != "OLD" {
		t.Errorf("history[0] (newest snapshot): got %+v want rev=2 code=OLD", got.History[0])
	}
	if got.History[1].Revision != 1 {
		t.Errorf("history[1]: got rev=%d want 1", got.History[1].Revision)
	}
}

// TestBumpStructureRevision_LegacyNormalised covers the case where the
// previous revision is missing (0) — bump must start from 1.
func TestBumpStructureRevision_LegacyNormalised(t *testing.T) {
	current := domain.Structure{ID: "st-x", Code: "X", Name: "X"}
	draft := domain.Structure{Code: "X2", Name: "X2"}
	got := BumpStructureRevision(current, draft)
	if got.Revision != 2 {
		t.Errorf("Revision from legacy: got %d want 2", got.Revision)
	}
	if len(got.History) != 1 || got.History[0].Revision != 1 {
		t.Errorf("History should contain normalised rev 1 snapshot, got %+v", got.History)
	}
}

func TestResolveStructureRevision(t *testing.T) {
	st := domain.Structure{
		ID: "st-1", Code: "CUR", Name: "Current",
		Revision: 3,
		History: []domain.StructureRevision{
			{Revision: 2, Code: "v2", Name: "Two"},
			{Revision: 1, Code: "v1", Name: "One"},
		},
	}

	t.Run("nil pin returns current", func(t *testing.T) {
		got, err := ResolveStructureRevision(st, nil)
		if err != nil {
			t.Fatal(err)
		}
		if got.Revision != 3 || got.Code != "CUR" {
			t.Errorf("got %+v want current", got)
		}
	})

	t.Run("pin == current returns current", func(t *testing.T) {
		cur := 3
		got, err := ResolveStructureRevision(st, &cur)
		if err != nil {
			t.Fatal(err)
		}
		if got.Revision != 3 {
			t.Errorf("got rev %d want 3", got.Revision)
		}
	})

	t.Run("pin in history returns frozen snapshot", func(t *testing.T) {
		pin := 1
		got, err := ResolveStructureRevision(st, &pin)
		if err != nil {
			t.Fatal(err)
		}
		if got.Revision != 1 || got.Code != "v1" || got.Name != "One" {
			t.Errorf("got %+v want rev=1/code=v1/name=One", got)
		}
	})

	t.Run("unknown pin returns StructureRevisionError with context", func(t *testing.T) {
		pin := 42
		_, err := ResolveStructureRevision(st, &pin)
		if err == nil {
			t.Fatal("expected error for unknown pin")
		}
		var sre *StructureRevisionError
		if !errors.As(err, &sre) {
			t.Fatalf("expected *StructureRevisionError, got %T (%v)", err, err)
		}
		if sre.Pin != 42 || sre.CurrentRevision != 3 || sre.StructureID != "st-1" {
			t.Errorf("error context wrong: %+v", sre)
		}
		// available revisions must be sorted ascending and include current + history.
		want := []int{1, 2, 3}
		if !reflect.DeepEqual(sre.AvailableRevisions, want) {
			t.Errorf("available: got %v want %v", sre.AvailableRevisions, want)
		}
	})
}

// TestResolveStructureForPin_ReifiesStructure checks the returned Structure is
// shape-complete and uses the snapshot fields for BOM resolution.
func TestResolveStructureForPin_ReifiesStructure(t *testing.T) {
	st := domain.Structure{
		ID: "st-1", Code: "CUR", Name: "Current", WidthMm: 600, HeightMm: 720, DepthMm: 560,
		Revision: 2,
		Components: []domain.ComponentInstance{{ComponentID: "c-cur", Quantity: 2}},
		History: []domain.StructureRevision{
			{
				Revision: 1, Code: "v1", Name: "One",
				Components: []domain.ComponentInstance{{ComponentID: "c-v1", Quantity: 5}},
			},
		},
	}
	pin := 1
	got, err := ResolveStructureForPin(st, &pin)
	if err != nil {
		t.Fatal(err)
	}
	if got.Revision != 1 || got.Code != "v1" {
		t.Errorf("revision/code: got %d/%q want 1/v1", got.Revision, got.Code)
	}
	if len(got.Components) != 1 || got.Components[0].ComponentID != "c-v1" || got.Components[0].Quantity != 5 {
		t.Errorf("components not reified from snapshot: %+v", got.Components)
	}
	// Shape-complete: id carried over even though it does not affect BOM.
	if got.ID != "st-1" {
		t.Errorf("ID should be carried over: got %q", got.ID)
	}
}

// TestCaptureProjectItemStructurePins covers the close-time pinning and the
// no-op paths (items without structure / structure missing from catalog).
func TestCaptureProjectItemStructurePins(t *testing.T) {
	st := domain.Structure{ID: "st-1", Revision: 4}
	modWithStructure := domain.Module{ID: "mod-st", StructureID: "st-1"}
	modFlat := domain.Module{ID: "mod-flat"} // no structure
	modMissingStructure := domain.Module{ID: "mod-missing", StructureID: "st-ghost"}

	catalog := domain.Catalog{
		Modules:    []domain.Module{modWithStructure, modFlat, modMissingStructure},
		Structures: []domain.Structure{st},
	}

	items := []domain.ProjectItem{
		{ID: "i1", ModuleID: "mod-st"},
		{ID: "i2", ModuleID: "mod-flat"},
		{ID: "i3", ModuleID: "mod-missing"},
		{ID: "i4", ModuleID: "mod-unknown"}, // module not in catalog
		{ID: "i5", ModuleID: "mod-st", StructureRevisionPin: intPtr(1)}, // stale pin should be overwritten
	}

	got := CaptureProjectItemStructurePins(items, catalog)

	if got[0].StructureRevisionPin == nil || *got[0].StructureRevisionPin != 4 {
		t.Errorf("i1 (mod-st): pin = %v want 4", pinVal(got[0].StructureRevisionPin))
	}
	if got[1].StructureRevisionPin != nil {
		t.Errorf("i2 (flat module): pin should be nil, got %v", pinVal(got[1].StructureRevisionPin))
	}
	if got[2].StructureRevisionPin != nil {
		t.Errorf("i3 (missing structure): pin should be nil, got %v", pinVal(got[2].StructureRevisionPin))
	}
	if got[3].StructureRevisionPin != nil {
		t.Errorf("i4 (unknown module): pin should be nil, got %v", pinVal(got[3].StructureRevisionPin))
	}
	if got[4].StructureRevisionPin == nil || *got[4].StructureRevisionPin != 4 {
		t.Errorf("i5 (stale pin): should be overwritten to 4, got %v", pinVal(got[4].StructureRevisionPin))
	}
}

func intPtr(v int) *int { return &v }

func pinVal(p *int) interface{} {
	if p == nil {
		return nil
	}
	return *p
}

File: packages/domain/src/structures/versioning.test.ts
/**
 * Tests for structure revision versioning (#108 / Slice 1).
 * One describe per logical unit, following conventions.md.
 */

import { describe, expect, it } from 'vitest';
import type {
  Catalog,
  Module,
  ProjectItem,
  Structure,
} from '../types';
import { ResolutionError } from '../errors';
import {
  bumpStructureRevision,
  captureProjectItemStructurePins,
  DEFAULT_STRUCTURE_REVISION,
  resolveStructureForPin,
  resolveStructureRevision,
  snapshotStructureRevision,
  structureRevision,
} from './versioning';

const baseStructure: Structure = {
  id: 'struct-1',
  code: 'EST-1',
  name: 'Cuerpo 1',
  revision: 1,
  externalDims: { width: 600, height: 720, depth: 560 },
  presets: [
    { id: 'p-600', name: '600', width: 600, height: 720, depth: 560 },
  ],
  components: [
    { componentId: 'comp-side', quantity: 2 },
    { componentId: 'comp-base', quantity: 1 },
  ],
  active: true,
};

describe('bumpStructureRevision', () => {
  it('increments revision 1 → 2 and reports the old revision', () => {
    const { structure, oldRevision } = bumpStructureRevision(baseStructure, {
      code: 'EST-1',
      name: 'Cuerpo 1 v2',
      externalDims: { width: 600, height: 720, depth: 560 },
      components: [{ componentId: 'comp-side', quantity: 3 }],
      active: true,
    });
    expect(structure.revision).toBe(2);
    expect(oldRevision).toBe(1);
  });

  it('snapshots the previous revision into history[0] with components intact', () => {
    const { structure } = bumpStructureRevision(baseStructure, {
      code: 'EST-1',
      name: 'Cuerpo 1 v2',
      components: [{ componentId: 'comp-side', quantity: 3 }],
      active: true,
    });
    expect(structure.history).toHaveLength(1);
    const snap = structure.history?.[0];
    expect(snap?.revision).toBe(1);
    expect(snap?.code).toBe('EST-1');
    expect(snap?.name).toBe('Cuerpo 1');
    // Original component list preserved verbatim.
    expect(snap?.components).toEqual([
      { componentId: 'comp-side', quantity: 2 },
      { componentId: 'comp-base', quantity: 1 },
    ]);
    expect(snap?.presets).toEqual(baseStructure.presets);
    expect(snap?.externalDims).toEqual(baseStructure.externalDims);
  });

  it('second edit produces revision 3 and accumulates history (length 2, newest-first)', () => {
    const first = bumpStructureRevision(baseStructure, {
      code: 'EST-1',
      name: 'v2',
      components: [{ componentId: 'comp-side', quantity: 3 }],
      active: true,
    }).structure;
    const second = bumpStructureRevision(first, {
      code: 'EST-1',
      name: 'v3',
      components: [{ componentId: 'comp-side', quantity: 4 }],
      active: true,
    }).structure;
    expect(second.revision).toBe(3);
    expect(second.history).toHaveLength(2);
    // Newest-first: history[0] is the just-superseded revision 2.
    expect(second.history![0]!.revision).toBe(2);
    expect(second.history![1]!.revision).toBe(1);
  });

  it('preserves id and ignores any revision/history coming from the draft', () => {
    const { structure } = bumpStructureRevision(baseStructure, {
      // draft attempts to smuggle revision/history — must be ignored.
      id: 'should-be-ignored',
      code: 'EST-1',
      name: 'v2',
      revision: 99,
      history: [{ revision: 50, code: 'x', name: 'x' }],
      components: [],
      active: true,
    } as unknown as Omit<Structure, 'id' | 'revision' | 'history'>);
    expect(structure.id).toBe('struct-1');
    expect(structure.revision).toBe(2);
    expect(structure.history).toHaveLength(1);
    expect(structure.history![0]!.revision).toBe(1);
  });

  it('treats missing revision as 1 (legacy data) when computing oldRevision', () => {
    const legacy: Structure = { ...baseStructure, revision: undefined };
    const { structure, oldRevision } = bumpStructureRevision(legacy, {
      code: 'EST-1',
      name: 'v2',
      components: [],
      active: true,
    });
    expect(oldRevision).toBe(1);
    expect(structure.revision).toBe(2);
    expect(structure.history![0]!.revision).toBe(1);
  });
});

describe('resolveStructureRevision', () => {
  // Build a structure at revision 3 with rev 1 and rev 2 in history.
  const atRev3: Structure = bumpStructureRevision(
    bumpStructureRevision(baseStructure, {
      code: 'EST-1',
      name: 'v2',
      components: [{ componentId: 'comp-side', quantity: 3 }],
      active: true,
    }).structure,
    {
      code: 'EST-1',
      name: 'v3',
      components: [{ componentId: 'comp-side', quantity: 4 }],
      active: true,
    },
  ).structure;

  it('pin undefined → returns the current (live) revision', () => {
    const r = resolveStructureRevision(atRev3, undefined);
    expect(r.revision).toBe(3);
    expect(r.code).toBe('EST-1');
    expect(r.name).toBe('v3');
    expect(r.components).toEqual([{ componentId: 'comp-side', quantity: 4 }]);
  });

  it('pin === current revision → returns the current revision', () => {
    const r = resolveStructureRevision(atRev3, 3);
    expect(r.revision).toBe(3);
    expect(r.components).toEqual([{ componentId: 'comp-side', quantity: 4 }]);
  });

  it('pin found in history → returns that frozen revision verbatim', () => {
    const r1 = resolveStructureRevision(atRev3, 1);
    expect(r1.revision).toBe(1);
    expect(r1.name).toBe('Cuerpo 1');
    expect(r1.components).toEqual([
      { componentId: 'comp-side', quantity: 2 },
      { componentId: 'comp-base', quantity: 1 },
    ]);
    const r2 = resolveStructureRevision(atRev3, 2);
    expect(r2.revision).toBe(2);
    expect(r2.name).toBe('v2');
    expect(r2.components).toEqual([{ componentId: 'comp-side', quantity: 3 }]);
  });

  it('unknown pin throws ResolutionError with structureId, pin, currentRevision, availableRevisions', () => {
    try {
      resolveStructureRevision(atRev3, 99);
      throw new Error('should have thrown');
    } catch (err) {
      expect(err).toBeInstanceOf(ResolutionError);
      const ctx = (err as ResolutionError).context as Record<string, unknown>;
      expect(ctx.structureId).toBe('struct-1');
      expect(ctx.pin).toBe(99);
      expect(ctx.currentRevision).toBe(3);
      expect(ctx.availableRevisions).toEqual([1, 2, 3]);
      expect(ctx.field).toBe('structureRevisionPin');
    }
  });

  it('resolveStructureForPin reifies into a full Structure usable by resolveBom', () => {
    const reified = resolveStructureForPin(atRev3, 1);
    expect(reified.id).toBe('struct-1');
    expect(reified.revision).toBe(1);
    expect(reified.components).toEqual([
      { componentId: 'comp-side', quantity: 2 },
      { componentId: 'comp-base', quantity: 1 },
    ]);
    // Live (no pin) reification also works.
    const live = resolveStructureForPin(atRev3);
    expect(live.revision).toBe(3);
  });
});

describe('snapshotStructureRevision / structureRevision / DEFAULT_STRUCTURE_REVISION', () => {
  it('snapshot only captures BOM-relevant fields (drops notes/active/history)', () => {
    const s: Structure = {
      ...baseStructure,
      notes: 'a note',
      history: [{ revision: 0, code: 'x', name: 'x' }],
    };
    const snap = snapshotStructureRevision(s);
    expect(snap).toEqual({
      revision: 1,
      code: 'EST-1',
      name: 'Cuerpo 1',
      externalDims: baseStructure.externalDims,
      presets: baseStructure.presets,
      components: baseStructure.components,
    });
    // No leakage of non-BOM fields:
    expect((snap as unknown as Record<string, unknown>).notes).toBeUndefined();
    expect((snap as unknown as Record<string, unknown>).active).toBeUndefined();
  });

  it('structureRevision normalizes missing revision to DEFAULT_STRUCTURE_REVISION', () => {
    expect(DEFAULT_STRUCTURE_REVISION).toBe(1);
    expect(structureRevision(baseStructure)).toBe(1);
    expect(structureRevision({ ...baseStructure, revision: undefined })).toBe(1);
    expect(structureRevision({ ...baseStructure, revision: 7 })).toBe(7);
  });
});

describe('captureProjectItemStructurePins', () => {
  const moduleWithStructure: Module = {
    id: 'mod-1',
    code: 'MOD-1',
    name: 'M1',
    structureId: 'struct-1',
    hardwareLines: [],
  };
  const moduleWithoutStructure: Module = {
    id: 'mod-2',
    code: 'MOD-2',
    name: 'M2',
    hardwareLines: [],
  };
  const moduleMissingStructure: Module = {
    id: 'mod-3',
    code: 'MOD-3',
    name: 'M3',
    structureId: 'struct-missing',
    hardwareLines: [],
  };

  const catalog: Catalog = {
    materials: [],
    edges: [],
    hardware: [],
    optionGroups: [],
    modules: [moduleWithStructure, moduleWithoutStructure, moduleMissingStructure],
    structures: [
      { ...baseStructure, revision: 5 },
    ],
  };

  const items: ProjectItem[] = [
    { id: 'i1', moduleId: 'mod-1', quantity: 1, optionChoices: {} },
    { id: 'i2', moduleId: 'mod-2', quantity: 1, optionChoices: {} },
    { id: 'i3', moduleId: 'mod-3', quantity: 1, optionChoices: {} },
  ];

  it('pins items to their module structure revision; leaves others untouched', () => {
    const pinned = captureProjectItemStructurePins(items, catalog);
    expect(pinned[0]?.structureRevisionPin).toBe(5); // has structure → pinned
    expect(pinned[1]?.structureRevisionPin).toBeUndefined(); // no structureId → unchanged
    expect(pinned[2]?.structureRevisionPin).toBeUndefined(); // structure missing → unchanged
    // Original items not mutated.
    expect(items[0]?.structureRevisionPin).toBeUndefined();
  });

  it('overwrites a stale pin on re-capture (idempotent at a given catalog)', () => {
    const stale: ProjectItem[] = [
      { id: 'i1', moduleId: 'mod-1', quantity: 1, optionChoices: {}, structureRevisionPin: 999 },
    ];
    const pinned = captureProjectItemStructurePins(stale, catalog);
    expect(pinned[0]?.structureRevisionPin).toBe(5);
  });
});
